synapse/federation/federation_server.py

```diff
@@ -160,6 +166,43 @@ def __init__(self, hs):
             )
             return 400, response
 
+        # We process PDUs and EDUs in parallel. This is important as we don't
+        # want to block things like to device messages from reaching clients
+        # behind the pontentially expensive handling of PDUs.
+        pdu_results, _ = await make_deferred_yieldable(
+            defer.gatherResults(
+                [
+                    run_in_background(
+                        self._handle_pdus_in_txn, origin, transaction, request_time
+                    ),
+                    run_in_background(self._handle_edus_in_txn, origin, transaction),
+                ],
+                consumeErrors=True,
+            ).addErrback(unwrapFirstError)
```

Comment:
I'm not entirely sure how the error handling is working here. Will this raise an exception if one of the wrapped methods raise?

Reply:
This is basically a copy from: https://github.com/matrix-org/synapse/blob/master/synapse/util/async_helpers.py#L167-L186

The way `defer.gatherResults` works is that:

1. `consumeErrors=True` means that it won't report errors to the default "unhandled error" logging
2. `gatherResults` will raise a `FirstError` that wrapse the actual exception. This is a bit annoying as we might be attempting to handle particular exceptions further up the stack, so we use `unwrapFirstError` which does what it says on the tin.